client/app/react-components/QueryViewTop.js

```diff
@@ -0,0 +1,238 @@
+    const toastMessages = {
+      success: 'Query saved',
+      error: 'Query could not be saved',
+      archiveError: 'Query could not be archived',
+    };
+    return (
+      <div className="query-page-wrapper">
+        {canEdit ? <AlertUnsavedChanges isDirty={this.isDirty()} onChangeLocation={this.onChangeLocation} /> : null}
+        {this.state.toast && <div id="toast-container" className="toast-bottom-right"><ToastMessageAnimated type={this.state.toast} message={toastMessages[this.state.toast]} /></div>}
+        <QueryViewHeader
+          canEdit={canEdit}
+          query={query}
+          updateQuery={this.updateAndSaveQuery}
+          currentUser={this.props.currentUser}
+          hasDataSources={dataSources.length > 0}
+          dataSource={dataSource}
+          sourceMode={this.props.sourceMode}
+          showPermissionsControl={this.props.clientConfig.showPermissionsControl}
+          duplicateQuery={this.duplicateQuery}
+          archiveQuery={this.archiveQuery}
+          getTags={this.props.getTags}
+          tags={this.props.tags}
+          clientConfig={this.props.clientConfig}
+          Events={this.props.Events}
+        />
+        <QueryViewMain
+          clientConfig={this.props.clientConfig}
+          canEdit={canEdit}
+          currentUser={this.props.currentUser}
+          basePath={this.props.basePath}
+          baseQuery={query}
+          queryResult={this.props.queryResult}
+          updateAndSaveQuery={this.updateAndSaveQuery}
+          isDirty={this.isDirty()}
+          dataSource={dataSource}
+          dataSources={dataSources}
+          setDataSource={this.setDataSource}
+          sourceMode={this.props.sourceMode}
+          executeQuery={this.props.executeQuery}
+          executeQueryResponse={this.props.executeQueryResponse}
+          updateQuery={this.updateQuery}
+          KeyboardShortcuts={this.props.KeyboardShortcuts}
+        />
+      </div>
+    );
+  }
+}
+
+function fetchQuery(props) {
+  if (props.queryId) {
+    return {
+      query: {
+        url: `${props.clientConfig.basePath}api/queries/${props.queryId}`,
+        andThen: query => ({
+          queryResult: query.latest_query_data_id ? {
+            url: `${props.clientConfig.basePath}api/query_results/${query.latest_query_data_id}`,
+          } : undefined,
+
+          dataSources: {
+            url: `${props.clientConfig.basePath}api/data_sources`,
+            then: dataSources => ({
+              value: dataSources.filter(dataSource =>
+                (!dataSource.viewOnly || dataSource.id === query.data_source_id)),
+            }),
+          },
+        }),
+      },
+      saveQuery: newQuery => ({
+        query: { value: newQuery },
+        saveQueryResponse: {
+          force: true,
+          url: `${props.clientConfig.basePath}api/queries/${props.queryId}`,
+          method: 'POST',
+          body: JSON.stringify(newQuery),
+        },
+      }),
```

Comment:
@openjck the problem I'm facing is that saving a query produces a query value in response with an updated `version` field; the server checks that `version` increases in each subsequent save to prevent conflicts. Should I just remove my `query` field here and rename `saveQueryResponse` to `query` if I want to update the query prop from the server each time it's saved?

Reply:
I tried this (renaming `saveQueryResponse` to `query`) and it solves the immediate issue of updating `query` to match the server response. The problem now is that if there's a failure to save I'd like to keep the old value of `query` rather than having no data to render. Do I have to move `query` into `state` to accomplish this?